File: semiconductor-stock/scripts/generate-stock-json.mjs
/**
 * GitHub Actions / 로컬: 일봉 → public/data/*.json
 * 우선순위: 한국투자증권(KIS) Open API → Yahoo Finance 폴백
 *
 * 환경변수:
 *   KIS_APP_KEY, KIS_APP_SECRET  (필수, KIS 사용 시)
 *   KIS_ENV=vps                  (모의투자 API 서버)
 */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import {
  fetchSixMonthBarsWithLive,
  hasKisCredentials,
  yahooSymbolToKisCode,
} from "./kis-client.mjs";

const SYMBOLS = ["005930.KS", "000660.KS"];
const root = path.dirname(fileURLToPath(import.meta.url));
const outDir = path.join(root, "..", "public", "data");

function chartUrl(symbol) {
  return (
    `https://query1.finance.yahoo.com/v8/finance/chart/${encodeURIComponent(symbol)}` +
    "?interval=1d&range=6mo"
  );
}

async function fetchBarsYahoo(symbol) {
  const res = await fetch(chartUrl(symbol), {
    headers: { "User-Agent": "Mozilla/5.0 (compatible; SemiconductorStock/1.0)" },
  });
  if (!res.ok) throw new Error(`${symbol} Yahoo ${res.status}`);
  const json = await res.json();
  const result = json.chart?.result?.[0];
  const timestamps = result?.timestamp ?? [];
  const quote = result?.indicators?.quote?.[0];
  const meta = result?.meta;
  const bars = [];

  for (let i = 0; i < timestamps.length; i++) {
    const open = quote?.open?.[i];
    const high = quote?.high?.[i];
    const low = quote?.low?.[i];
    const close = quote?.close?.[i];
    if (open == null || high == null || low == null || close == null) continue;
    const ts = timestamps[i];
    bars.push({
      date: new Intl.DateTimeFormat("en-CA", {
        timeZone: "Asia/Seoul",
        year: "numeric",
        month: "2-digit",
        day: "2-digit",
      }).format(new Date(ts * 1000)),
      open,
      high,
      low,
      close,
      volume: quote?.volume?.[i] ?? 0,
    });
  }
  if (bars.length < 30) throw new Error(`${symbol} 데이터 부족`);

  if (meta?.regularMarketPrice != null && bars.length) {
    const livePrice = meta.regularMarketPrice;
    const liveDate = meta.regularMarketTime
      ? new Intl.DateTimeFormat("en-CA", {
          timeZone: "Asia/Seoul",
          year: "numeric",
          month: "2-digit",
          day: "2-digit",
        }).format(new Date(meta.regularMarketTime * 1000))
      : null;
    const last = bars[bars.length - 1];
    if (liveDate && last.date < liveDate) {
      bars.push({
        date: liveDate,
        open: livePrice,
        high: livePrice,
        low: livePrice,
        close: livePrice,
        volume: 0,
      });
    } else {
      last.close = livePrice;
      last.high = Math.max(last.high, livePrice);
      last.low = Math.min(last.low, livePrice);
    }
  }

  return bars;
}

async function fetchBars(symbol) {
  if (hasKisCredentials()) {
    const code = yahooSymbolToKisCode(symbol);
    const bars = await fetchSixMonthBarsWithLive(code);
    return { bars, source: "kis" };
  }

  console.warn(
    `[warn] KIS 키 없음 → Yahoo 폴백 (${symbol}). ` +
      "GitHub Secrets에 KIS_APP_KEY / KIS_APP_SECRET 을 등록하세요."
  );
  const bars = await fetchBarsYahoo(symbol);
  return { bars, source: "yahoo" };
}

fs.mkdirSync(outDir, { recursive: true });

const sourceLabel = hasKisCredentials() ? "kis" : "yahoo-fallback";
console.log(`Baking semiconductor stock JSON via ${sourceLabel}…`);

for (const symbol of SYMBOLS) {
  const file = path.join(outDir, `${symbol}.json`);
  try {
    const { bars, source } = await fetchBars(symbol);
    fs.writeFileSync(
      file,
      JSON.stringify({
        symbol,
        source,
        fetchedAt: new Date().toISOString(),
        bars,
      })
    );
    const last = bars[bars.length - 1];
    console.log(
      `Wrote ${file} (${bars.length} bars, last=${last?.date} close=${last?.close}, source=${source})`
    );
  } catch (e) {
    if (fs.existsSync(file)) {
      console.warn(
        `[warn] ${symbol} bake 실패 — 기존 JSON 유지:`,
        e instanceof Error ? e.message : e
      );
      continue;
    }
    throw e;
  }
}

File: semiconductor-stock/scripts/generate-stock-json.py
#!/usr/bin/env python3
"""
삼성전자·SK하이닉스 일봉 → public/data/*.json
PyKRX (KRX 공개 시세) — API 키 불필요
"""
from __future__ import annotations

import json
import sys
import time
from datetime import datetime, timedelta
from pathlib import Path
from zoneinfo import ZoneInfo

try:
    from pykrx import stock
except ImportError:
    print("pykrx 미설치: pip install pykrx", file=sys.stderr)
    sys.exit(1)

KST = ZoneInfo("Asia/Seoul")
ROOT = Path(__file__).resolve().parent.parent
OUT_DIR = ROOT / "public" / "data"

# symbol(웹 키) → 종목코드
SYMBOLS = [
    ("005930.KS", "005930"),
    ("000660.KS", "000660"),
]


def kst_today() -> datetime:
    return datetime.now(KST)


def ymd(d: datetime) -> str:
    return d.strftime("%Y%m%d")


def iso_date(d) -> str:
    if hasattr(d, "strftime"):
        return d.strftime("%Y-%m-%d")
    return str(d)[:10]


def fetch_bars(code: str, months: int = 7) -> list[dict]:
    end = kst_today()
    start = end - timedelta(days=months * 31)
    df = stock.get_market_ohlcv_by_date(ymd(start), ymd(end), code)
    if df is None or df.empty:
        raise RuntimeError(f"{code}: PyKRX 일봉 비어 있음")

    # 컬럼: 시가 고가 저가 종가 거래량 (버전별 영문일 수 있음)
    colmap = {}
    for c in df.columns:
        name = str(c)
        if name in ("시가", "Open", "open"):
            colmap["open"] = c
        elif name in ("고가", "High", "high"):
            colmap["high"] = c
        elif name in ("저가", "Low", "low"):
            colmap["low"] = c
        elif name in ("종가", "Close", "close"):
            colmap["close"] = c
        elif name in ("거래량", "Volume", "volume"):
            colmap["volume"] = c

    required = ("open", "high", "low", "close")
    if not all(k in colmap for k in required):
        raise RuntimeError(f"{code}: 예상 컬럼 없음 {list(df.columns)}")

    bars: list[dict] = []
    for idx, row in df.iterrows():
        try:
            open_ = float(row[colmap["open"]])
            high = float(row[colmap["high"]])
            low = float(row[colmap["low"]])
            close = float(row[colmap["close"]])
        except (TypeError, ValueError):
            continue
        if any(v != v for v in (open_, high, low, close)):  # NaN
            continue
        vol = 0.0
        if "volume" in colmap:
            try:
                vol = float(row[colmap["volume"]])
                if vol != vol:
                    vol = 0.0
            except (TypeError, ValueError):
                vol = 0.0
        bars.append(
            {
                "date": iso_date(idx),
                "open": open_,
                "high": high,
                "low": low,
                "close": close,
                "volume": vol,
            }
        )

    bars.sort(key=lambda b: b["date"])
    if len(bars) < 30:
        raise RuntimeError(f"{code}: 일봉 부족 ({len(bars)}건)")
    return bars


def write_symbol(symbol: str, code: str) -> None:
    out = OUT_DIR / f"{symbol}.json"
    try:
        bars = fetch_bars(code)
        payload = {
            "symbol": symbol,
            "source": "pykrx",
            "fetchedAt": datetime.now(KST).isoformat(),
            "bars": bars,
        }
        out.write_text(
            json.dumps(payload, ensure_ascii=False, separators=(",", ":")),
            encoding="utf-8",
        )
        last = bars[-1]
        print(
            f"Wrote {out} ({len(bars)} bars, last={last['date']} "
            f"close={last['close']}, source=pykrx)"
        )
    except Exception as e:
        if out.exists():
            print(f"[warn] {symbol} bake 실패 — 기존 JSON 유지: {e}", file=sys.stderr)
            return
        raise


def main() -> None:
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    print("Baking semiconductor stock JSON via PyKRX…")
    for i, (symbol, code) in enumerate(SYMBOLS):
        write_symbol(symbol, code)
        if i < len(SYMBOLS) - 1:
            time.sleep(0.8)  # KRX 과도 요청 완화


if __name__ == "__main__":
    main()
